Network/MainServer.py

```diff
@@ -0,0 +1,258 @@
+            ((player2.x + player2.width, player2.y), (player2.x + player2.width, player2.y + player2.height)), # player2 back
+            ((player2.x, player2.y), (player2.x, player2.y + player2.height)),  # player2 front
+        ]
+        vertices = [
+            (self.x, self.y), 
+            (self.x + self.width, self.y),
+            (self.x, self.y + self.height), 
+            (self.x + self.width, self.y + self.height)
+        ]
+        for face in faces:
+            for vertice in vertices:
+                intersection = self.line_intersection(previous_pos, vertice, face[0], face[1])
+                if intersection is not None:
+                    break
+            if intersection:
+                if face == faces[0]:
+                    self.x, self.y = intersection[0] + 1, intersection[1]
+                    relative_y = (self.y + (self.width / 2)) - player1.y                  
+                    deflection_weight = relative_y / player1.height * 2 - 1
+                    angle_offset = deflection_weight * 45
+                    self.direction =  angle_offset + 90
+                elif face == faces[1]:
+                    self.x, self.y = intersection[0] + player1.width + 1, intersection[1]
+                    relative_y = (self.y + (self.width / 2)) - player1.y                   
+                    deflection_weight = relative_y / player1.height * 2 - 1
+                    angle_offset = deflection_weight * 45
+                    self.direction =  angle_offset + 90
+                elif face == faces[2]:
+                    self.x, self.y = intersection[0] - self.width - player2.width  - 1, intersection[1]
+                    relative_y = (self.y + (self.width / 2)) - player2.y                   
+                    deflection_weight = relative_y / player1.height * 2 - 1
+                    angle_offset = deflection_weight * 45
+                    self.direction =  - angle_offset - 90 
+                else:
+                    self.x, self.y = intersection[0] - self.width - 1, intersection[1]
+                    relative_y = (self.y + (self.width / 2)) - player2.y
+                    deflection_weight = relative_y / player1.height * 2 - 1
+                    angle_offset = deflection_weight * 45
+                    self.direction =  - angle_offset - 90
+                player1.speed += 0.125
+                player2.speed += 0.125
+                player1.speed = min(player1.speed, 5)
+                player2.speed = min(player2.speed, 5)
+                self.speed += 0.25
+                self.speed = min(9, self.speed)
+
+        if self.y <= 0 or self.y + self.height >= 400:
+            self.direction = 180 - self.direction
+            if self.y <= 0:
+                self.y = 1
+            else:
+                self.y = 400 - self.height - 1
+
+        # point detection
+        if self.x + self.width <= 0 or self.x >= 800:
```

Comment:
Nitpick: These position checks are duplicated, is it worth pulling these out into variables? This also flattens the control flow, for example:
```py
player1_scored = self.x >= 800
player2_scored = self.x + self.width <= 0

if player1_scored:
    player1.score += 1

if player2_scored:
    player2.score += 1

if player1_scored or player2_scored:
    self.x, self.y = self.starting_pos
    self.direction = random.randint(1, 4) * 90 + 45
    self.speed = 5
    player1.speed = 3
    player2.speed = 3
```